### ISSUE TITLE
reproduce paper with pytorch

### ISSUE THREAD

**Issue by ZeR0ll** [jirenz/HiDDeN#3]
hi I am interested about your works and try to reproduce paper with pytorch, because I know nothing about torch and lua so it's hard for me to know the detail about your code.I have some problem in reproduce the paper:
I have some difference in implementation,I train in 1 128x128 image with totol random 30 bit watermark and want to make it learn how to embed watermark in that image.I put adversary after the noise layer.and use sigmoid instead of linear layer in adversary.
1.what is the purpose of Lg(I_en) = log(1-A(I_en)),actually it is -log(1-A(I_en)) right? I think this loss try to train encoder to make discriminator misclassify the encoded image to cover img,I only know little about GAN,it is that right?
2.after test I found my training loss of message decrease to a small number but have high error rate in test,about 11 bits wrong in 30 bit watermark.do you have some idea about the reason?I train and test it without noise attack.

I will be grateful if you can give me some advice,I am only an undergraduate student and I don't have profound understanding about deep learning.thanks very much!

**Comment by jirenz** [comment]
Hi ZeROII. Thank you for your interest.
* The adversary should use un-perturbed encoded image because we don't want to cripple the adversary. 
* You are correct about what L_g is trying to do. The sign is correct in the paper. Because the loss of the adversary is log(A(I_en)) + log(1 - A(I_co)). 
* For debugging purposes, I suggest not using the adversary and using a slightly larger coefficient for l_2 loss on image difference. The adversary only becomes necessary when you try to train networks robust to noise.
* How many training images did you use? What is the message error rate on the training images? I would suggest looking at bugs in the code as we didn't run into any trouble with generalization.
Feel free to ask more questions.

**Comment by ZeR0ll** [comment]
thanks for your reply, it is really helpful.I find bugs in my code so the generalization was not a problem.I use only 1 image and embed random watermark during training,it can converge quickly.
And have you notice that the model use high GPU memory when the watermark length is large,for example 1024 bits,I think the concatenated layer in Eocoder's parameters might be the reason.

**Comment by jirenz** [comment]
* I agree that it can happen. (although not necessarily because when concatenating, you can fake the big replicated block of message bits by setting stride etc. and the actual memory consumption won't go up) Still the current network architecture and encoding scheme is not designed for large L because every bit is encoded everywhere on the output image and 1024 is way too much for that.
* If you want to have high capacity, you can train a small message length encoder on a small image and work tile by tile for big images (as is done in the paper).
* If you want the model to encode the watermark end to end, you can break up the message as part of the encoder and only put a piece of the message to the feature map of a specific location (or better, can the network learn to do location dependent encoding?).
